### ISSUE TITLE
`needless_return` does not lint on `async` function

### ISSUE THREAD

**Issue by minseongg** [rust-lang/rust-clippy#7042]
Clippy produces a false negative on a `needless_return` when there is an unneeded `return` statement in async function.
This is happening on both stable and nightly rust

I tried this code:

```rust
async fn test() -> Result<(), ()> {
    return Ok(());
}

fn main() -> Result<(), ()> {
    futures::executor::block_on(test())
}
```

I expected to have Clippy inform me about having an unneeded `return`, with this warning.

```
warning: unneeded `return` statement
 --> src/main.rs:2:5
  |
2 |     return Ok(());
  |     ^^^^^^^^^^^^^^ help: remove `return`: `Ok(())`
  |
  = note: `#[warn(clippy::needless_return)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return
```

Instead, this happened:

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
```

### Meta

- `cargo clippy -V`: `clippy 0.1.52 (07e0e2ec 2021-03-24)`
- `rustc -Vv`:
```
rustc 1.53.0-nightly (07e0e2ec2 2021-03-24)
binary: rustc
commit-hash: 07e0e2ec268c140e607e1ac7f49f145612d0f597
commit-date: 2021-03-24
host: x86_64-apple-darwin
release: 1.53.0-nightly
LLVM version: 12.0.0
```
